src/Microsoft.AspNetCore.Razor.Evolution/CodeGeneration/DesignTimeCSharpRenderer.cs

```diff
@@ -38,14 +33,17 @@ public override void VisitCSharpExpression(CSharpExpressionIRNode node)
 
                     for (var i = 0; i < node.Children.Count; i++)
                     {
-                        var childNode = node.Children[i];
-
-                        if (childNode is CSharpTokenIRNode)
+                        var token = node.Children[i] as RazorIRToken;
+                        if (token != null && token.IsCSharp)
                         {
-                            AddLineMappingFor(childNode);
+                            AddLineMappingFor(token);
+                            Context.Writer.Write(token.Content);
+                        }
+                        else
+                        {
+                            // There may be something else inside the expression like a Template or another extension node.
+                            Visit(node.Children[i]);
```

Comment:
This looks like our opportunity to tighten up the set of allowed/valid nodes. For instance a template is implemented as an extension node, but it's expected to appear in an expession context.